### PR TITLE
pullsync: coalesce batch writes

### PR DESCRIPTION
**Pull request by acud**
This should also improve the CPU usage, since leveldb doesn't need to do individual writes. The downside is that because it might take longer to do the actual write, we might end up wanting the same chunk from multiple peers since this creates a possible data race (we want the chunk from one peer which is slow, and another peer which is fast, resulting in deliveries from both).

Open for discussion about this. Not entirely sure if it's needed if we get the other change which buffers the triggers to the subscriptions, which should solve the problem.

### CODE REVIEW

**Review by ldeffenb — COMMENTED** [review]
If there was an error reading the delivery causing the for loop to terminate, but it was after at least one chunk was added to chunksToPut, then the storage.Put is going to hide the read error by resetting err.

**Review by zelig — COMMENTED** [review]
improve error handling

**Review by janos — APPROVED** [review]
LGTM. It might be more clear with some appropriately named functions, but it is well commented, so maybe some other time. 👍

File: pkg/pullsync/pullsync.go [hunk 1]
```diff
@@ -202,33 +202,53 @@ func (s *Syncer) SyncInterval(ctx context.Context, peer swarm.Address, bin uint8
...
 			return 0, ru.Ruid, swarm.ErrInvalidChunk
 		}
-
-		if err = s.storage.Put(ctx, storage.ModePutSync, chunk); err != nil {
-			return 0, ru.Ruid, fmt.Errorf("delivery put: %w", err)
+		chunksToPut = append(chunksToPut, chunk)
+	}
+	if len(chunksToPut) > 0 {
+		s.metrics.DbOpsCounter.Inc()
+		if ierr := s.storage.Put(ctx, storage.ModePutSync, chunksToPut...); ierr != nil {
```
Comment (zelig) [inline pkg/pullsync/pullsync.go:239]: this is a bit convoluted, but ok

**Review by zelig — APPROVED** [review]

**Pull request merged**